domain/infosync/info.go

```diff
@@ -314,22 +317,26 @@ func (is *InfoSyncer) getServerInfoByID(ctx context.Context, id string) (*Server
 	return info, nil
 }
 
+// global variables only used in failpoint tests.
+var (
+	allServerInfo4Test map[string]*ServerInfo
+	serverInfo4Test    *ServerInfo
+)
+
+// SetAllServerInfo4Test set the value of `allServerInfo4Test` used in unit test.
+func SetAllServerInfo4Test(infos map[string]*ServerInfo) {
+	allServerInfo4Test = infos
+}
+
+// SetServerInfo4Test set the value of `serverInfo4Test` used in unit test.
+func SetServerInfo4Test(info *ServerInfo) {
+	serverInfo4Test = info
+}
+
 // GetAllServerInfo gets all servers static information from etcd.
 func GetAllServerInfo(ctx context.Context) (map[string]*ServerInfo, error) {
 	failpoint.Inject("mockGetAllServerInfo", func() {
-		res := map[string]*ServerInfo{
-			"fa598405-a08e-4e74-83ff-75c30b1daedc": {
-				Labels: map[string]string{
-					"zone": "zone1",
-				},
-			},
-			"ad84dbbd-5a50-4742-a73c-4f674d41d4bd": {
-				Labels: map[string]string{
-					"zone": "zone2",
-				},
-			},
-		}
-		failpoint.Return(res, nil)
+		failpoint.Return(allServerInfo4Test, nil)
```

Comment:
Maybe we can use `failpoint.Enable("github.com/pingcap/tidb/domain/infosync/mockGetServerInfo", fmt.Sprintf("return(\"%s\")", the_result_of_json_marshaled_info))`, then you could unmarshal that string in `failpoint.Inject("mockGetAllServerInfo", ` without global vars.

Reply:
Is this the recommended way to do so. This will decrease the readablity as some value are relatively long.

Reply:
You could construct a struct and marshal it into string. Yeah, this is actually a best practice of failpoint, though lots of people don't know how to pass value into the inject point.

Reply:
done. @xhebox  PTAL again, thank you~